### PR TITLE
[C++] Small fixes for c++

### LINKED ISSUE #146: [C++] Test failure in DatasetReadingTest.DatasetRead
This test needs to be made idempotent.

```
5/5 Test #4: DatasetReadingTest.DatasetRead ...........***Failed    0.05 sec
Note: Google Test filter = DatasetReadingTest.DatasetRead
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DatasetReadingTest
[ RUN      ] DatasetReadingTest.DatasetRead
/home/lidavidm/Code/upstream/arrow-cookbook/cpp/code/datasets.cc:56: Failure
Failed
'arrow::dataset::FileSystemDataset::Write(write_options, std::move(scanner))' failed with Invalid: Could not write to /tmp/cookbook_cpp_airquality as the directory is not empty and existing_data_behavior is to error
[  FAILED  ] DatasetReadingTest.DatasetRead (21 ms)
[----------] 1 test from DatasetReadingTest (21 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (21 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DatasetReadingTest.DatasetRead

 1 FAILED TEST
```

### PR DESCRIPTION
**Pull request by wjones127**
## Summary
 * Remove UseAsync since it was removed in Arrow 7.0 ([commit](https://github.com/apache/arrow/commit/2e8b8365d602d8cb14a37ea4f349824549e28e8f))
 * Change existing data behavior to make the datasets setup idempotent. Fixes #146.
 * Added explicit overrides to appease `-Winconsistent-missing-override`
 * Added `ARROW_RETURN_NOT_OK` to appease `-Wunused-result`

### CODE REVIEW

**Comment by wjones127** [comment]
Can I update this to Arrow 7.0.0?

Not sure how the environment file is updated. Was going to install it, do `conda install arrow-cpp==7.0.0`, and export that environment, but I'm not able to create the current environment locally. Is it possible that it's somehow not portable? I'm on an M1 Mac. cc @lidavidm @westonpace

```
❯ conda env create --file cpp/environment.yml
Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound: 
  - cryptography==3.4.7=py39hbca0aa6_0
  - gxx_linux-64==11.1.0=h33c4e4b_0
  - pysocks==1.7.1=py39hf3d152e_3
  - openssl==1.1.1l=h7f98852_0
  - libthrift==0.15.0=he6d91bd_1
  - python==3.9.6=h49503c6_1_cpython
  - libgfortran-ng==11.1.0=h69a702a_8
  - docutils==0.17.1=py39hf3d152e_0
  - gcc_linux-64==11.1.0=h97fdae6_0
  - libgcc-devel_linux-64==11.1.0=h80e7780_8
  - cmake==3.21.1=h8897547_0
  - rhash==1.4.1=h7f98852_0
  - grpc-cpp==1.42.0=ha1441d3_1
  - libffi==3.3=h58526e2_2
  - libssh2==1.9.0=ha56f1ee_6
  - libstdcxx-ng==11.2.0=he4da1e4_12
  - libuv==1.42.0=h7f98852_0
  - libzlib==1.2.11=h36c2ea0_1013
  - libprotobuf==3.19.4=h780b84a_0
  - ca-certificates==2021.10.8=ha878542_0
  - binutils_linux-64==2.36=hf3e587d_0
  - make==4.3=hd18ef5c_1
  - _openmp_mutex==4.5=1_gnu
  - aws-c-cal==0.5.11=h95a6274_0
  - libbrotlienc==1.0.9=h7f98852_5
  - gcc_impl_linux-64==11.1.0=h6b5115b_8
  - libevent==2.1.10=hcdb4288_3
  - abseil-cpp==20210324.2=h9c3ff4c_0
  - libedit==3.1.20191231=he28a2e2_2
  - libbrotlidec==1.0.9=h7f98852_5
  - ncurses==6.2=h58526e2_4
  - krb5==1.19.2=hcc1bbae_0
  - libstdcxx-devel_linux-64==11.1.0=h80e7780_8
  - gxx_impl_linux-64==11.1.0=h6b5115b_8
  - zlib==1.2.11=h36c2ea0_1013
  - zstd==1.5.2=ha95c52a_0
  - libgomp==11.2.0=h1d223b6_12
  - libcurl==7.78.0=h2574ce0_0
  - binutils_impl_linux-64==2.36.1=h193b22a_2
  - expat==2.4.1=h9c3ff4c_0
  - setuptools==49.6.0=py39hf3d152e_3
  - arrow-cpp==6.0.1=py39h01fd06f_8_cpu
  - libbrotlicommon==1.0.9=h7f98852_5
  - glog==0.5.0=h48cff8f_0
  - libsanitizer==11.1.0=h56837e0_8
  - libutf8proc==2.7.0=h7f98852_0
  - markupsafe==2.0.1=py39h3811e60_0
  - libev==4.33=h516909a_1
  - cffi==1.14.6=py39he32792d_0
  - chardet==4.0.0=py39hf3d152e_1
  - certifi==2021.10.8=py39hf3d152e_1
  - sqlite==3.36.0=h9cd32fc_0
  - orc==1.7.2=h1be678f_0
  - re2==2021.11.01=h9c3ff4c_0
  - gflags==2.2.2=he1b5a44_1004
  - _libgcc_mutex==0.1=conda_forge
  - clang-11==11.1.0=default_ha53f305_1
  - numpy==1.21.1=py39hdbf815f_0
  - lz4-c==1.9.3=h9c3ff4c_1
  - libllvm11==11.1.0=hf817b99_2
  - c-ares==1.18.1=h7f98852_0
  - ld_impl_linux-64==2.36.1=hea4e1c9_2
  - libgcc-ng==11.2.0=h1d223b6_12
  - readline==8.1=h46c0cb4_0
  - aws-c-io==0.10.5=hfb6a706_0
  - tk==8.6.10=h21135ba_1
  - libnghttp2==1.43.0=h812cca2_0
  - xz==5.2.5=h516909a_1
  - libgfortran5==11.1.0=h6c583b3_8
  - s2n==1.0.10=h9b69904_0
  - aws-c-common==0.6.2=h7f98852_0
  - snappy==1.1.8=he1b5a44_3
  - pyarrow==6.0.1=py39hff6fa39_8_cpu
  - aws-checksums==0.1.11=ha31a3da_7
  - clangxx==11.1.0=default_ha53f305_1
  - bzip2==1.0.8=h7f98852_4
  - aws-sdk-cpp==1.8.186=hb4091e7_3
  - libclang-cpp11.1==11.1.0=default_ha53f305_1
  - gtest==1.10.0=h4bd325d_7
  - clang==11.1.0=ha770c72_1
  - brotlipy==0.7.0=py39h3811e60_1001
  - libopenblas==0.3.17=pthreads_h8fe5266_1
  - aws-c-event-stream==0.2.7=h3541f99_13
  - clang-tools==11.1.0=default_ha53f305_1
 ```

**Comment by lidavidm** [comment]
The environment appears to pin all the packages.

So I would instead create a fresh environment and dump it to environment.yml afterwards.

**Comment by westonpace** [comment]
Well, actually, I think the environment-specific point might be a good one.  I'm pretty sure my system is very close to the CI system so I get away with it.  Hmm...maybe it's time to properly adopt conda lockfiles instead.

File: cpp/CONTRIBUTING.md [hunk 1]
```diff
@@ -120,7 +120,22 @@ If you are using conda then there is file `cpp/requirements.yml` which can be
...
+```
+conda create -n cookbook-cpp --file cpp/conda-osx-arm64.lock
+```
+
+To update dependencies modify `cpp/requirements.yml` and then run
+
+```
+cd cpp
+conda-lock --file environment.yml --kind explicit --platform osx-arm64
+conda-lock -f environment.yml -p linux-aarch64 -p linux-64 -p osx-arm64
```
[the remark this answers is not in the document]
Comment (wjones127) [reply cpp/CONTRIBUTING.md:138]: Whoops. Yeah I hadn't save my latest update to the file.

[the remark this answers is not in the document]
Comment (wjones127) [reply cpp/CONTRIBUTING.md]: Though Marius' suggestion was to use the first one, as I understand it.

**Review by westonpace — APPROVED** [review]
Thanks for cleaning things up and figuring out conda locks.

**Pull request merged**